### ISSUE TITLE
1.1.333, 1.1.334 regression - high CPU usage (~100% single core) in VS Code

### ISSUE THREAD

**Issue by tomasz-c** [microsoft/pyright#6314]
**Describe the bug**

- update the pyrigtht extension to version 1.1.333 or 1.1.334,
- open the directory containing the git repository and python scripts in VS Code
- click on any python script to display it in VS Code,

This creates an electron process that consumes ~100% of the CPU (on one core). Then pyright doesn't work (it doesn't suggest the code). Closing VS Code does not end this process, it stays and consumes CPU.
If a directory without a git repository is opened, or the git extension in VS Code is disabled ("git.enabled": false), pyright works correctly. However, the cause seems to be pyright, as downgrading to 1.1.332 solves this problem.

**VS Code extension or command-line**
The issue occurs as an extension in VS Code
Tested in VS Code versions: 1.83.1 and 1.84.0 on Linux
I also have "python.languageServer" set to "None", but changing it to 'jedi' doesn't seem to change anything.

**Comment by erictraut** [comment]
Thanks for reporting the issue, but you haven't provided any repro steps. Without a repro case, I won't be able to make any further progress on the issue. Is your code base public? If so, please provide the steps required to clone the repo and install the dependencies. If your code isn't public, you'll need to narrow down the problem to a particular piece of code and provide a minimal sample or give me temporary access to a repo that contains the relevant portions.

**Comment by mephinet** [comment]
I can confirm the issue, but I've failed to single out a minimal setup that shows the issue in a reproducible way...

**Comment by erictraut** [comment]
I've looked through all of the changes between 1.1.333 and 1.1.334. The one that stands out as a potential culprit is this the bug fix for #6234. Another possibility is the fix for #6224.

I still need a repro.

**Comment by tomasz-c** [comment]
I thought this happens with any repository and python code, but it doesn't.
This can be reproduced with sample public code:

https://github.com/uwi-info3180/flask-sqlite

Open the flask-sqlite directory as 'Open folder' in VS Code and click on the 'run.py' file to load into the editor.

**Comment by erictraut** [comment]
Hmm, I'm not able to repro the problem you're seeing either with or without the Git extension enabled. It's curious that you're able to repro it only with the Git extension disabled. That seems to indicate some interaction between Pyright and the Git extension, but I don't have any working theories.

@debonte, @rchiodo — let me know if you have any ideas.

**Comment by mephinet** [comment]
I can fully confirm @tomasz-c analysis. I cloned the flask-sqlite, created a venv, pip-installed all requirements, and deactivated all plugins except for pyright. I'm running code 1.83.1 and python 3.11.5, bot installed via the official repo of Arch Linux.
Pyright 1.1.334: as soon as I open e.g. run.py, an electron process is spawned that consumed 100% cpu and won't get killed when I close vscode. When I strace the process, all I see are futex/mmap/madvise/mprotect calls.
Downgrading to  1.1.333 results in identical behavior.
Downgrading to 1.1.332 solves the issue, pyright works as expected again.
I've attached the logs of the 1.1.332 and the 1.1.334 run - they don's show any significant differences, to my eyes... 
[logs.zip](https://github.com/microsoft/pyright/files/13257656/logs.zip)

**Comment by erictraut** [comment]
@mephinet, thanks for the info. Can you confirm that you also see the problem go away if you disable the Git extension (i.e. set `git.enabled` to "false")?

**Comment by erictraut** [comment]
It would also be interesting to see if the problem can be repro'ed using the command-line version of pyright. Does it also hang?

**Comment by mephinet** [comment]
Yes, I can confirm that the issue goes away when disabling the git extension 😑

Running pyright 1.1.334 on the commandline results with argument "." take 2.0s and produces identical output (errors in models.py and views.py)

When checking the diff between 1.1.332 and 1.1.333, a change that I find interesting is https://github.com/microsoft/pyright/pull/6231 which changes seems to move some analysis into the background, introduces a background waiter, and fixes some thread shutdown code..? Maybe someone with access to the pyrx repo can shed some light to this...

**Comment by tomasz-c** [comment]
I also checked the official VS Code compilation from https://code.visualstudio.com/ and the problem also exists, but then the CPU-loading process is called 'code' and it also remains after closing the application.

I am attaching the result of the command `timeout 1 strace --timestamps=format:unix,precision:us -T -s 500 -p 18424 -o strace.txt`, i.e. a one-second strace for the burdening process. Maybe it will be helpful for someone.
[strace.txt](https://github.com/microsoft/pyright/files/13257771/strace.txt)

**Comment by mephinet** [comment]
I have a new observation I want to share. When I said, I'm using python 3.11.5 provided by Arch, this wasn't the whole story. I've installed https://github.com/pyenv/pyenv - and even though I haven't enabled it for the repository I'm using to reproduce this issue, it still means that the pyenv shim is executed.

For those who don't know pyenv: it injects a shell script named "python" (as well as python3, pip, and many others) into your path so that whenever python is executed, this script can check whether you're in a directory where a specific python version is configured. If so, it uses this locally compiled version instead of the system-wide installation. If there is no such configuration, it falls back to /usr/bin/python.

I just gave it a few tries - and reproducably:
When running vscode with the pyenv shim, I run into our issue.
When setting PATH to only /usr/local/bin:/usr/bin:/bin:/usr/local/sbin and then executing vscode, I'm not able to reproduce this issue.

Can someone please try to reproduce this? I'm still not fully trusting my eyes here. (And still have no clue about the role of the git extension)

@tomasz-c Do you have pyenv installed too?

**Comment by tomasz-c** [comment]
> @tomasz-c Do you have pyenv installed too?

I also use Arch Linux and Python version 3.11.5 but I don't have pyenv installed.
Setting PATH to '/usr/local/bin:/usr/bin:/bin:/usr/local/sbin' doesn't change anything (problem still exists)

I tried to manually set the interpreter in VS Code to python 3.10.13 (manually compiled) and the problem still exists.

**Comment by mephinet** [comment]
Thanks - on my second notebook with "identical" setup, I'm also not able to reproduce this relevance of the PATH - so this was probably a coincidence...

**Comment by lpz-stratslab** [comment]
Just as a data point, I think I experience the same issue for `pyright 1.1.334`, but it can actually take up more than one core (two in my case):
```
$ ps aux -q 601714
USER         PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
lpzcore+  601714  174  2.2 943904 367688 ?       Rl   09:18 365:19 /opt/code-server-v4.18.0-linux-x86_64/lib/node --max-old-space-size=3072 /home/workspaces/lpz/.ide/data/code-server/extensions/ms-pyright.pyright-1.1.334/dist/server.js --cancellationReceive=file:7e619b1ed44f5d4d83bf9f66d7c77ea64141fd06f0 --node-ipc --clientProcessId=256
```
It's actually [code-server](https://github.com/coder/code-server/tree/v4.18.0) which is based on `Code 1.83.1` and looks like also susceptible to the bug. OS is `Ubuntu 20.04.6 LTS` and Python is not the distro one, but a custom one in `3.9.10` version.

**Comment by heejaechang** [comment]
> When checking the diff between 1.1.332 and 1.1.333, a change that I find interesting is https://github.com/microsoft/pyright/pull/6231 which changes seems to move some analysis into the background, introduces a background waiter, and fixes some thread shutdown code..? Maybe someone with access to the pyrx repo can shed some light to this...

I looked the PR and most of them should only affect `pylance` not `pyright`. the `analyzeFile` stuff `move some analysis into the background` seems referring to, only `pylance` use that, so `pyright` shouldn't affect by them.

by the way, if CPU is consumed 100% and if the issue is indeed with pylance/pyright, then it sounds like there are some code path where cancellation is not working as expected (either not passed down, or not checked)

**Comment by erictraut** [comment]
@jfcherng, do you have a set of repro steps? I haven't been able to repro this problem yet. Without a repro case, I won't be able to make any progress on it.

**Comment by tomasz-c** [comment]
It seems that version 1.1.335 solved this problem (accidentally?). Version 1.1.336 also works fine.
Going back to version 1.1.334 I can still reproduce this problem, which suggests that the problem was on the pyright side.

Maybe someone else could confirm that the problem disappeared in version 1.1.335? @mephinet @lpz-stratslab @jfcherng
I think that if someone else confirms this, we can close this issue.

**Comment by jfcherng** [comment]
Not my case but I am on Sublime Text though. Still checking whether this is a pyright issue or not.

**Comment by rchl** [comment]
I don't know if it's the same issue but in Sublime Text this happens for unsaved buffers.

Unsaved buffers in ST use a `buffer:<number>` URI (at least in not-yet-released build).

The process locks up while trying to traverse this URI.

I've done a small patch to pyright to log stuff in `getCompletionSuggestions`:

```diff
diff --git a/packages/pyright-internal/src/analyzer/importResolver.ts b/packages/pyright-internal/src/analyzer/importResolver.ts
index 23ac4a5fd..5f65d0923 100644
--- a/packages/pyright-internal/src/analyzer/importResolver.ts
+++ b/packages/pyright-internal/src/analyzer/importResolver.ts
@@ -187,8 +187,17 @@ export class ImportResolver {
         const root = this.getParentImportResolutionRoot(sourceFilePath, execEnv.root);
         const origin = ensureTrailingDirectorySeparator(getDirectoryPath(sourceFilePath));

+        this.serviceProvider.console().info(`sourceFilePath: ${sourceFilePath}`)
+
+        let c = 100
         let current = origin;
+
         while (this._shouldWalkUp(current, root, execEnv)) {
+            this.serviceProvider.console().info(`current: ${current}, root: ${root}`)
+            c--
+            if (c <= 0) {
+                break
+            }
             this._getCompletionSuggestionsAbsolute(
                 sourceFilePath,
                 execEnv,
```

Then opened a new buffer:

```py
import 
```

and triggered completion after `"import "`.

The logs said:

```
:: [09:05:13.581] --> LSP-pyright textDocument/completion (2): {'textDocument': {'uri': 'buffer:587'}, 'position': {'line': 0, 'character': 7}}
LSP-pyright: sourceFilePath: buffer:587
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
LSP-pyright: current: buffer:/, root: buffer:/
```

**Comment by erictraut** [comment]
@rchiodo, given that you're working on a major change to the way that pyright tracks paths/uris, you may be interested in the comment above ^^^. Perhaps your new code will address the issue with uris that start with "buffer:"?

**Comment by rchl** [comment]
I would even say that any non-file scheme should more or less be treated the same (unless there are some exceptions). For example VSCode uses `untitled:` for unsaved buffers and I would expect that `untitled:`, `buffer:` or `foo:` should all have the same handling.

**Comment by rchiodo** [comment]
The new change should handle this case, as long as the URIs follow the same scheme as outlined here: https://github.com/microsoft/vscode-uri. We're using that to parse URIs.

**Comment by erictraut** [comment]
Do you think that the problem reported above will go away with your change? In other words, did you run across any bugs in the previous code that could explain a hang when these `untitled:` URIs are used?

**Comment by rchl** [comment]
The URIs we are using follow the URI specification so will work with `vscode-uri`.

Note that URI can either have or not have the `authority` component. So for example `buffer:foo` has no `authority` and  has path `foo` while `buffer://foo/bar` has `authority` `foo` and path `bar`. Not sure if that's also something relevant to your changes.

**Comment by debonte** [comment]
This all reminds me of https://github.com/microsoft/pyright/issues/6232. Maybe that fix has regressed somehow?

**Comment by rchiodo** [comment]
The new URI changes won't have such a problem.

**Comment by mephinet** [comment]
I can confirm that v1.1.336 solved the issue for me, in VS Code.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.337, which I just published. It will also be included in a future release of pylance.
